### PR TITLE
First version of the merge functionality.

### PR DESCRIPTION
**Pull request by thomasmauerhofer**
This is the first version to merge two projects.

Testrun: https://jenkins.catrob.at/view/All-Categories/view/Catroid-multi-job/job/Catroid-Multi-Job-Custom-Branch-RELOADED/1791/

### CODE REVIEW

File: catroid/src/org/catrobat/catroid/ProjectManager.java [hunk 1]
```diff
@@ -504,6 +517,374 @@ private void correctAllNestedReferences(Script script) {
 		}
 	}
 
+	public void mergeProjectInCurrentProject(String projectNameToMergeFrom, FragmentActivity activity) {
+
```
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/ProjectManager.java]: Remove all those unnecessary newlines after a method declaration to be consistent. Especially with *your own* code :wink:

**Comment by robertpainsi** [comment]
There is a lot of new code in the ProjectManager which should certainly be moved to it's own class(es)

File: catroidTest/src/org/catrobat/catroid/uitest/content/interaction/MergeProjectTest.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+import org.catrobat.catroid.content.bricks.WaitBrick;
+import org.catrobat.catroid.content.bricks.WhenStartedBrick;
+import org.catrobat.catroid.io.StorageHandler;
+import org.catrobat.catroid.ui.MainMenuActivity;
+import org.catrobat.catroid.uitest.util.BaseActivityInstrumentationTestCase;
+
+public class MergeProjectTest extends BaseActivityInstrumentationTestCase<MainMenuActivity> {
+
+	private String projectNameTo = "mergeTo";
+
```
Comment (robertpainsi) [inline catroidTest/src/org/catrobat/catroid/uitest/content/interaction/MergeProjectTest.java]: Remove-a me

File: catroid/src/org/catrobat/catroid/utils/Utils.java [hunk 3]
```diff
@@ -502,4 +512,6 @@ public static void setSelectAllActionModeButtonVisibility(View selectAllActionMo
 			selectAllActionModeButton.setVisibility(View.GONE);
 		}
 	}
+
+
```
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/utils/Utils.java]: Remove us all

**Comment by robertpainsi** [comment]
Please take a look at your code again and remove all unnecessary newlines you can find. :+1:

**Comment by robertpainsi** [comment]
The german translation should be removed because it's handled by crowdin, I guess?
ping @thmq

**Comment by thmq** [comment]
just so!

File: catroid/src/org/catrobat/catroid/utils/UtilMerge.java [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+		// copy SpriteListofLists
+		projectToMergeInto = copySpriteListOfLists(projectToMergeInto, projectToMergeFrom);
+
+		// copy UserBrickVariables
+		projectToMergeInto = copyUserBrickVariables(projectToMergeInto, projectToMergeFrom);
+
+		return projectToMergeInto;
+	}
+
+	private static Project copySkripts(Project projectToMergeInto, Project projectToMergeFrom) {
```
Comment (avihric) [inline catroid/src/org/catrobat/catroid/utils/UtilMerge.java]: should be `copyScripts`

**Comment by avihric** [comment]
Why are lines 282-297 in `UtilMerge.java` commented out?

**Comment by avihric** [comment]
Also your commit msg does not meet the guidelines. Please update PR.

File: catroid/src/org/catrobat/catroid/formulaeditor/DataContainer.java [hunk 5]
```diff
@@ -296,6 +303,13 @@ public UserList addProjectUserList(String userListName) {
 		return userListToAdd;
 	}
 
+	public void setProjectLists(DataContainer datacontainer) {
+
```
Comment (avihric) [inline catroid/src/org/catrobat/catroid/formulaeditor/DataContainer.java]: remove the blank line

File: catroid/src/org/catrobat/catroid/formulaeditor/DataContainer.java [hunk 6]
```diff
@@ -366,4 +388,12 @@ public UserList getUserList() {
 		}
 		return null;
 	}
+
+	public void setUserBrickVariables(DataContainer containerFrom) {
+
```
Comment (avihric) [inline catroid/src/org/catrobat/catroid/formulaeditor/DataContainer.java]: repeat previous action :P

File: catroid/src/org/catrobat/catroid/formulaeditor/DataContainer.java [hunk 7]
```diff
@@ -366,4 +388,12 @@ public UserList getUserList() {
...
+		for (int key : containerFrom.userBrickVariables.keySet()) {
+				addUserBrickVariable(key, containerFrom.userBrickVariables.get(key));
+
```
Comment (avihric) [inline catroid/src/org/catrobat/catroid/formulaeditor/DataContainer.java]: you know what to do :)

File: catroid/src/org/catrobat/catroid/utils/UtilMerge.java [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+
+	private static final String TAG = UtilMerge.class.getSimpleName();
+
+	private static Project project;
+
+	public static void mergeProjectInCurrentProject(String projectNameToMergeFrom, FragmentActivity activity) {
+		try {
+			project = ProjectManager.getInstance().getCurrentProject();
+			Project projectToMergeFrom = loadProjectContent(projectNameToMergeFrom, activity);
+
```
Comment (avihric) [inline catroid/src/org/catrobat/catroid/utils/UtilMerge.java]: remove this blank line

File: catroid/src/org/catrobat/catroid/utils/UtilMerge.java [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+
+			if (!checkHeader(project.getXmlHeader(), projectToMergeFrom.getXmlHeader()) && project.getSpriteList().size() > 1) {
+				showDifferentResolutionDialog(projectToMergeFrom, activity);
+			} else {
+				if (!checkMergeConflicts(project, projectToMergeFrom, activity)) {
+
```
Comment (avihric) [inline catroid/src/org/catrobat/catroid/utils/UtilMerge.java]: same here

**Comment by avihric** [comment]
Please remove all comments and blank lines (at least use a consistent way)

**Comment by thmq** [comment]
this pr can be closed, has been replaced by #1260 (rebased and squashed)

**Comment by avihric** [comment]
I have merged #1263 because in your PR there were still 2 unsolved new lines.

**Pull request closed without merging**